addons/website_sale_coupon/controllers/main.py

```diff
@@ -28,6 +28,14 @@ def cart(self, **post):
         order.recompute_coupon_lines()
         return super(WebsiteSale, self).cart(**post)
 
+    @http.route(['/coupon/<string:code>'], type='http', auth='public', website=True, sitemap=False)
+    def activate_coupon(self, code, **post):
+        request.session['pending_coupon'] = {'code': code}
```

Comment:
The dict question is about the specs (it keeps the message for later re-use), which makes it hard to resolve without @sbuhl. Personally, I don't agree, because I like the fact that the error message is kept, and I find it does add to the user-friendliness.

About `_apply_on_order`, it still works in the same way. The `_apply_on_order` didn't support general program `promo_code` rewards. To support this the whole `_apply_coupon` method is needed, which encapsulates this piece of logic just fine. So again, I'd say this is more about specs (do we want to support this for `promo_code`)... If, for some reason, we only wanted to support `coupon.coupon`, I'd have to separate the corresponding logic again from `apply_coupon`.

Reply:
> Personally, I don't agree, because I like the fact that the error message is kept, and I find it does add to the user-friendliness.

So, I apply a coupon that requires 1.000 min $ amount, I got the error for the rest of my life displayed on the website without any way to remove it? (I can't remove it since it was not applied so there is no reward line to remove, and the message is persistent in session so I can't alter that as an enduser)

I really doesn't feel right.

What's wrong with the technical approach I suggested (see below)? This is just technical, it doesn't impair the specs at all.
Actually there is no functional spec about the error message.

> - Access /coupon/123
> - The controller check if the code can be applied, if not, it store the code in session (for later reuse) and return the error message directly inside the redirect, something like request.redirect('/shop/cart?error_pending=xxx'). The error is not stored.
> - The /shop/cart template is in charge of displaying that url param if set (you can check how it is done with request.params.get('code_not_available') in the code
> - On recompute_coupon_lines or later naviguation, the error code is not shown anymore
>
> I think this is enough, the user saw the error once, he knows it. He can know:
>
> - Add stuff in his cart, the coupon will be added automatically
> - Reapplied the coupon by using the shared link again

In fine, the goal of the spec (the part related to saving the pending coupon if it can't be applied), is just to 'silently' reapplied it later if we can to ease the user experience. We shouldn't overthink it by saving an error etc, especially if it comes with complicated code.

Note: don't support promo_code, just coupon for now so it is ok

Reply:
> So, I apply a coupon that requires 1.000 min $ amount, I got the error for the rest of my life displayed on the website without any way to remove it? (I can't remove it since it was not applied so there is no reward line to remove, and the message is persistent in session so I can't alter that as an enduser)

This is a valid concern. However, I think it is rather minor, because, if you think about it, it doesn't affect the flow of the enduser (if he decides not to meet the condition, it is just a message to ignore). But it may be a good idea to remove the coupon from session on checkout for example. Also, it is only valid if we decide to not show anything at all, because otherwise the same problem would occur for the general message "This pending coupon xxx will be applied on checkout" (and I do think we should communicate the presence of the coupon to the enduser). And of course your session =/= the rest of your life.

So that leaves us with no message at all. Imagine an enduser applying a coupon, returning to the shop, only to think, how much products/value (whatever the condition) do I still need to add to the cart? The enduser probably decides to go back to the cart, to double-check, only to find the coupon is now gone (at least thinks it's gone). Worst-case the enduser abandons the cart if he can't find the link anymore.